src/pool.rs

```diff
@@ -603,19 +614,53 @@ impl ConnectionPool {
     /// Get a connection from the pool.
     pub async fn get(
         &self,
-        shard: usize,               // shard number
+        shard: Option<usize>,       // shard number
         role: Option<Role>,         // primary or replica
         client_stats: &ClientStats, // client id
     ) -> Result<(PooledConnection<'_, ServerPool>, Address), Error> {
-        let mut candidates: Vec<&Address> = self.addresses[shard]
-            .iter()
-            .filter(|address| address.role == role)
-            .collect();
+        let mut effective_shard_id: Option<usize> = shard;
+
+        // The base, unsharded case
+        if self.shards() == 1 {
+            effective_shard_id = Some(0);
+        }
+
+        let mut sort_by_error_count = false;
+        let mut candidates: Vec<_> = match effective_shard_id {
+            Some(shard_id) => self.addresses[shard_id].iter().collect(),
+            None => match self.settings.no_shard_specified_behavior {
+                NoShardSpecifiedHandling::Random => self.addresses.iter().flatten().collect(),
+                NoShardSpecifiedHandling::RandomHealthy => {
+                    sort_by_error_count = true;
+                    self.addresses.iter().flatten().collect()
+                }
+                NoShardSpecifiedHandling::Shard(shard) => {
+                    if shard >= self.shards() {
+                        return Err(Error::InvalidShardId(shard));
+                    } else {
+                        self.addresses[shard].iter().collect()
+                    }
+                }
+            },
+        };
+        candidates.retain(|address| address.role == role);
 
         // We shuffle even if least_outstanding_queries is used to avoid imbalance
         // in cases where all candidates have more or less the same number of outstanding
         // queries
         candidates.shuffle(&mut thread_rng());
+
+        // The branch should only be hit if no shard is specified and we are using
+        // random healthy routing mode
+        if sort_by_error_count {
```

Comment:
Should be either shuffle or sort, not both. The reader will be confused whether this is intentional or not, and if it was, it wouldn't be obvious as to why.

Reply:
I refactored the code to make it clear what we are doing.
So, we shuffle first because we want to randomize the selected instance even if we end up resorting (e.g. if the sorting metric had the same value for all instances)